storage/badger.go

```diff
@@ -0,0 +1,186 @@
+package storage
+
+import (
+	"github.com/dgraph-io/badger/v3"
+	"github.com/pkg/errors"
+	"os"
+)
+
+const (
+	// PerformanceMemTableSize is 3072 MB. Increases the maximum
+	// amount of data we can commit in a single transaction.
+	PerformanceMemTableSize = 3072 << 20
+
+	// PerformanceLogValueSize is 256 MB.
+	PerformanceLogValueSize = 256 << 20
+)
+
+type BadgerDatabase struct {
+	db            *badger.DB
+	opts          badger.Options
+	useWriteBatch bool
+}
+
+func NewBadgerDatabase(opts badger.Options, useWriteBatch bool) *BadgerDatabase {
+	return &BadgerDatabase{
+		db:            nil,
+		opts:          opts,
+		useWriteBatch: useWriteBatch,
+	}
+}
+
+func (bdb *BadgerDatabase) Setup() error {
+	db, err := badger.Open(bdb.opts)
+	if err != nil {
+		return errors.Wrapf(err, "Setup:")
+	}
+	bdb.db = db
+	return nil
+}
+
+func (bdb *BadgerDatabase) Update(fn func(Transaction) error) error {
+	var wb *badger.WriteBatch
+	if bdb.useWriteBatch {
+		wb = bdb.db.NewWriteBatch()
+	}
+
+	err := bdb.db.Update(func(txn *badger.Txn) error {
+		T := NewBadgerTransaction(txn, wb)
+		return fn(T)
+	})
+	if err != nil {
+		return errors.Wrapf(err, "Update:")
+	}
+
+	if bdb.useWriteBatch {
+		err = wb.Flush()
+		if err != nil {
+			return errors.Wrapf(err, "Update: Problem flushing write batch")
+		}
+		wb.Cancel()
+	}
+	return nil
+}
+
+func (bdb *BadgerDatabase) View(fn func(Transaction) error) error {
+	return bdb.db.View(func(txn *badger.Txn) error {
+		T := NewBadgerTransaction(txn, nil)
+		return fn(T)
+	})
+}
+
+func (bdb *BadgerDatabase) Close() error {
+	return bdb.db.Close()
+}
+
+func (bdb *BadgerDatabase) Erase() error {
+	return os.RemoveAll(bdb.opts.Dir)
```

Comment:
Would it be safer to use `bdb.db. DropAll()` instead? https://pkg.go.dev/github.com/dgraph-io/badger#DB.DropAll

Not sure what the intricacies are of how badger caches data in memory, and if manually deleting the data directory is the move. If `os.RemoveAll` is safer, it's worth adding a comment here on why we made the decision to use it here